### ISSUE TITLE
Rojo Headless Plugin API Proposal

### ISSUE THREAD

**Issue by boatbomber** [rojo-rbx/rojo#638]
Exposing a headless API into `_G` will be very valuable to the Rojo ecosystem. Users can make companion plugins, and we can even utilize it ourselves for one-time plugin injections for `rojo open`. See issues #321 and #305.
Those are use cases and requests, and we've established there is demand and utility. Now, this issue is attempting to flesh out an execution of this idea. This is an ongoing discussion, not set in stone. Tell us about your use cases and desires so we can make this as valuable as possible!

**Proposed interface:**
```Lua
_G.Rojo
```
Plugins and command bar all share the same _G which allows us to put our interface in. Using the key `"Rojo"` is easy to remember and follows branding.

**Functions:**
```Lua
Rojo:Connect(host: string?, port: string?): (boolean, string?)
```
Attempts to connect Rojo to the given host & port. Same behavior as the user clicking the Connect button (sync locking, etc, all behave the same). Returns a boolean of whether it connected or not, with a second return of a message if necessary.

```Lua
Rojo:Disconnect(): void
```
Attempts to disconnect any active sync session.

```Lua
Rojo:GetSetting(setting: string): any
```
Returns the value of the given setting.

```Lua
Rojo:SetSetting(setting: string, value: any): void
```
Sets the value of the given setting to the given value.

```Lua
Rojo:Notify(msg: string): void
```
Sends a Rojo notification. Will indicate on it that it comes from a third party plugin.

**Properties:** *(All read-only)*

```Lua
Rojo.Version: {number}
```
The Rojo plugin version. (Example: `{7, 2, 1}`)

```Lua
Rojo.ProtocolVersion: number
```
The Rojo plugin's protocol version.

```Lua
Rojo.Connected: boolean
```
Whether Rojo is currently connected to a serve session.

**Comment by boatbomber** [comment]
I aim to have a security permissions layer as well, much like Roblox's own permission system. When a third party plugin attempts to use an API for the first time you will need to approve it first. This will hopefully mitigate a lot of potential abuse.

**Comment by boatbomber** [comment]
I will also need to make it popup when something already exists at _G.Rojo, asking if you want this version to overwrite the existing version (will show you the versions and sources) so that you can have Rojo Boatly and Rojo simultaneously and control which one is exposing the API.

**Comment by Dekkonot** [comment]
This might seem obvious, but why not use a more dependable solution like a ModuleScript? Initial testing suggests that plugins share the state of ModuleScripts, so there doesn't seem to be a downside.

If we were to put a ModuleScript somewhere in the data model, consumers would be able to require it and then use the API from there. It relies less on busywaiting to ensure `_G.Rojo` is loaded and we could have its Archivable property set to `false` to make sure it doesn't get saved into a file.

The only downside I see is that we may introduce noise into someone's game, but I think I prefer that over use of `_G`?

I'm more than open to being told "that's unacceptable and I would hate it" but that's my opinion right now

**Comment by boatbomber** [comment]
What makes that more dependable? Just seems like more work and cases to consider since you start dealing with Instances and DataModel stuff instead of just Lua environments.

**Comment by Dekkonot** [comment]
The reason I say it seems more dependable is because it's ordered. You can have a consumer use `WaitForChild` to know when the Rojo API is ready rather than the snippet suggested in the [currently pending PR](https://github.com/boatbomber/rojo.space/blob/headless-api-docs/docs/headless-api.md):

```lua
local function GetRojo()
	local api = _G.Rojo
	while not api do
		task.wait()
		api = _G.Rojo
	end
	return api
end

local Rojo = GetRojo()
```

I think as a consumer I'd prefer it to be `local Rojo = require(Location:WaitForChild("RojoApi"))` but that's my opinion.

**Comment by boatbomber** [comment]
Yeah, that's a fair critique. I'm willing to rework it to be a ModuleScript if that's what people want.

**Comment by boatbomber** [comment]
Done. Updated the docs as well. The suggested snippet is now:
```Lua
local Rojo = require(game:WaitForChild("Rojo", math.huge))
print(Rojo.API.Version)
```
